src/frequenz/sdk/data_ingestion/resampling/individual_timeseries_resampler.py

```diff
@@ -0,0 +1,62 @@
+"""
+IndividualTimeSeriesResampler class for resampling individual metrics.
+
+Copyright
+Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+License
+MIT
+"""
+import logging
+from typing import Callable, Optional, Sequence
+
+from frequenz.sdk.core.timeseries import Sample
+from frequenz.sdk.data_ingestion.resampling.ring_buffer import RingBuffer
+
+logger = logging.Logger(__name__)
+
+MAX_BUFFER_SIZE = 1000
+
+ResamplingFunction = Callable[[Sequence[Sample], float], float]
+
+
+class IndividualTimeSeriesResampler:
+    """IndividualTimeSeriesResampler class."""
+
+    def __init__(  # pylint: disable=too-many-arguments
+        self,
+        buffer_size: int,
+        period_s: float,
+        max_data_age_in_periods: float,
+        resampling_function: ResamplingFunction,
+    ) -> None:
+        """Initialize the IndividualTimeSeriesResampler.
+
+        Args:
+            buffer_size: max number of samples to be stored in the buffer
+            period_s: resampling frequency, in seconds
+            max_data_age_in_periods: max age that samples shouldn't exceed in order
```

Comment:
@leandro-lucarella-frequenz could you please clarify what the `in_periods` was supposed to mean here?

Reply:
> . So if the resampling period is 0.2 seconds and max_data_age_in_periods is 5, then we'll consider data older than 1 second too old and ignore it.

@leandro-lucarella-frequenz In current implementation RingBuffer size is 1000 :( That would mean we won't remove data older then 1 second :( We will remove it when the size is > 1000.
@jakub-toptal  Could you change it?

Reply:
Yeah, as I said the idea was to be calculated from `max_data_age_in_periods` and `period_s`.

Reply:
Yes, I will change it. @leandro-lucarella-frequenz how exactly should it be calculated from these two? `max_data_age_in_periods / period_s`? I understand that `period_s` is the resampling window or the period after which next resampling round is triggered, but what does `max_data_age_in_periods` mean exactly?

Reply:
The calculation is lacking one key element, the input samples rate, sorry for overlooking that. So `max_data_age_in_periods` should also say which period to use. So what I had in mind is:

```py
buffer_len = resample_period_s / input_samples_period_s * max_age_in_resample_periods
```

So if the input sample period is 0.2s and resampling period is 1s and we want to consider only the last 3 resample periods (windows) as useful data, then each window will store 5 samples (`1/0.2`) and will we need `5*3 == 15` samples in the buffer at most.

Since the input sample rate will be calculated dynamically in the future, but we need to start with something, I suggest passing a `initial_input_samples_period_s` to the resampler, and use a default that is a typical sampling rate produced by a device (@tiyash-basu-frequenz maybe you have advice on what that value could be?). This would be a more future proof approach for when we calculate the input sample rate.

That said, I'm also fine to keep it simple for now and go with a `buffer_len` for this PR. We still need the `max_age_in_resample_periods` because when passing data to the resampling function we still need to look at the timestamps in the buffer and only pass the recent enough (relevant) samples to the resampling function (the buffer could still store very old data if there was some interruption for example).

Reply:
Renamed. As to `initial_input_samples_period_s` or `input_samples_period_s` I couldn't find them described in the issue so if any of these should be used, please let me know more precisely how.